Copy all information from thread into PEP (#9)
skirpichev/peps: peps/pep-9999.rst

PEP: 9999
Title: imath --- module for number-theoretic functions
Author: Neil Girdhar <misterhsheik@gmail.com>,
        Sergey B Kirpichev <skirpichev@gmail.com>,
Sponsor: TBD
Discussions-To: Pending
Status: Draft
Type: Standards Track
Created: 10-May-2025
Python-Version: 3.15
Post-History:
  02-Jun-2019,

>   `09-May-2025 <https://discuss.python.org/t/91337>`__,

Reviewer: ```suggestion
Post-History: 02-Jun-2019,
              `09-May-2025 <https://discuss.python.org/t/91337>`__,
```
Lets revert formatting changes.

Author: Done

Abstract
========

> This PEP proposes a new module for integer-related mathematical functions.

Reviewer: I doubt it's better.  Did you classify functions by domain?  by codomain?  Should math.floor() be moved?

Former variant clearly state module scope.  Maybe we can drop examples, but I don't think we should:
```suggestion
This PEP proposes new module for number-theoretic functions, like
:external+py3.14:func:`math.gcd` or :external+py3.14:func:`math.isqrt`.
```

Author: I asked Chat, and it said that things like combinatorics aren't part of number theory.  I think we can set the scope farther down.  How about "number theory, combinatorics, and integer representations."

I think we should drop the examples in the abstract.  The abstract should be as concise as possible.

Reviewer: Sorry, but I doubt we should care about opinion of hallucinating AI.

See e.g. the GMP manual:
* comb - mpz_bin_ui
* factorial - mpz_fac_ui
* gcd - mpz_gcd
* lcm - mpz_lcm

All in section 5.9 (Number Theoretic Functions).  The isqrt is in more fine-tuned section (root extraction).

Author: We don't need to care about AI, but we should find a good reference.  Maybe something a bit more authoritative than the GMP manual?  https://en.wikipedia.org/wiki/Combinatorics does not even mention number theory.  Do you have a citation that combinatorics is part of number theory?

Reviewer: > Maybe something a bit more authoritative than the GMP manual?

For example?

>  Do you have a citation that combinatorics is part of number theory?

I'm pretty sure that binomial coefficients are part of.

Author: According to Wikipedia: https://en.wikipedia.org/wiki/Mathematics#Areas_of_mathematics number theory and combinatorics are separate?  Do you read that differently?

> I'm pretty sure that binomial coefficients are part of.

Part of what?  I'd say they're part of combinatorics.  They're mentioned on the combinatorics page: https://en.wikipedia.org/wiki/Combinatorics

Motivation
==========

The :external+py3.14:mod:`math` module documentaton says: "This module provides

> access to the mathematical functions defined by the C standard."  Over time the

Reviewer: ```suggestion
access to the mathematical functions defined by the C standard."  But over time the
```

Author: Done.

module was populated with functions that aren't related to the C standard.

Also, the ``math`` module documentaton says: "Except when explicitly noted
otherwise, all return values are floats."  This is no longer true:  *None* of
``comb``, ``factorial``, ``gcd``, ``isqrt``, ``lcm``, ``perm``, ``floor``, or ``ceil`` returns
a float, but the docs don’t say so. In a hypothetical ``imath``, where they
belong, a similar catch-all: "Except when explicitly noted otherwise, all

> return values are ints." could tell the truth once.

Reviewer: ```suggestion
Also, the :external+py3.14:mod:`math` module documentation says: "Except when
explicitly noted otherwise, all return values are floats."  This is no longer
true: *None* of functions, listed in the "Number-theoretic functions" [1]_
subsection of the documentation returns a float, but the docs don’t say so.
In the proposed ``imath``, where they belong, a similar catch-all:
"Except when explicitly noted otherwise, all return values are integers." could
tell the truth once.
```
Lets be less verbose here and just add reference to docs.

Author: Done.

The ``imath`` module would parallel the :external+py3.14:mod:`math` module's
other siblings: :external+py3.14:mod:`cmath` and the
:external+py3.14:mod:`statistics`.  It would provide shared context, which
reduces verbosity in the documentaton and conceptual load.  It also aids
discoverability through grouping related functions and IDEs suggesting helpful

> completions.

Reviewer: ```suggestion
The ``imath`` module would be parallel the :external+py3.14:mod:`math` module's
other siblings: :external+py3.14:mod:`cmath` and the
:external+py3.14:mod:`statistics`.  It would provide shared context, which
reduces verbosity in the documentation and conceptual load.  It also aids
discoverability through grouping related functions and IDEs suggesting helpful
completions.
```
Spelling nits.

Author: Generally, you should prefer the active voice, so using parallel as a verb is better than "would be" (passive voice).

(Fixed the spelling of documentation everywhere.)

Specification
=============

The PEP proposes copying the follwing integer-related functions [1]_ to a new

> module, called ``imath``:

Reviewer: ```suggestion
The PEP proposes moving the following integer-related functions [1]_ to a new
module, called ``imath``:
```

Author: Do you think it will be confusing to use the word moving since we're leaving the old function pointers?   How about _exposing_?

Reviewer: The point that it's moving.  We just not deprecate old aliases.  Yet.

* ``gcd(a, b)``
* ``lcm(a, b)``
* ``isqrt(n)``
* ``factorial(n)``

> * ``comb(n, k)``

Reviewer: ```suggestion
* :external+py3.14:func:`~math.gcd`
* :external+py3.14:func:`~math.lcm`
* :external+py3.14:func:`~math.isqrt`
* :external+py3.14:func:`~math.factorial`
* :external+py3.14:func:`~math.comb`

Their aliases in :external+py3.14:mod:`math` will be :term:`soft deprecated`.
```

Author: Done.  The soft-deprecation is mentioned in the next section.

Author: I see,  you want to make the declaration of soft-deprecation here.  That makes sense.

Backwards Compatibility
=======================

Aliases in :external+py3.14:mod:`math` would be :term:`soft deprecated` [6]_,
that is, they would remain, but their use would be discouraged.  Thus, there

> are no anticipated code breaks.

Reviewer: ```suggestion
As aliases in :external+py3.14:mod:`math` will be kept for indefinite
time (their use would be discouraged), there are no anticipated code breaks.
```

Author: Done, but in two sentences.

Abstract
========

This PEP proposes a new module for mathematical functions related to number

> theory and combinatorics.

Reviewer: Could you argue why this is better?

Author: According to Wikipedia:
* combinatorics and number theory are different branches of mathematics, and
* functions we plan on exposing like the number of combinations (binomial coefficient) and permutations are part of combinatorics.

Therefore, I think our desired scope should be "number theory and combinatorics".

I also think that abstracts should be as concise as possible with details presented in papers/proposals.  So I prefer one sentence without examples.

Motivation
==========

The :external+py3.14:mod:`math` module documentation says: "This module provides
access to the mathematical functions defined by the C standard."  But, over time
the module was populated with functions that aren't related to those
mathematical functions defined in the C standard.

Also, the :external+py3.14:mod:`math` module documentation says: "Except when
explicitly noted otherwise, all return values are floats."  This is no longer
true:  *None* of the functions listed in the "Number-theoretic functions" [1]_
subsection of the documentation returns a float, but the documentation doesn't
say so. In the proposed ``imath`` module, where they belong, a similar
catch-all: "Except when explicitly noted otherwise, all return values are
integers." could tell the truth once.

The ``imath`` module would parallel the :external+py3.14:mod:`math` module's
other siblings: :external+py3.14:mod:`cmath` and the
:external+py3.14:mod:`statistics`.  It would provide shared context, which
reduces verbosity in the documentation and conceptual load.  It also aids
discoverability through grouping related functions and IDEs suggesting helpful

> completions.

Reviewer: Ditto.

Author: First paragraph:
* "as a state of the art" is being misused here, I'm not sure what it's trying to even express
* It's about discrepancy of the documentation with reality.  Since the documentation is about the C standard, there's no reason to also mention "returned values or accepted arguments"—that doesn't conflict with the topic.  Every paragraph should be focused.  If  you want to mention returned types, it should be in the second paragraph, which is about that.

Second para:
* "not longer true" -> "no longer true"
* I will remove ``imath`` etc. from the sentence (I like LHS better there)
* I will add some wording from the LHS
* "simplify description" -> "simplify the description"
* "of accepted arguments" -> "of the accepted arguments"
* "for both the...module" -> "for functions in both"

Third para:
* "Apparently" is a very odd way to word this.  Apparently to whom, and why?  I you want to say this, it needs to be reworded.
* "Let's make same" is grammatically incorrect.

Specification
=============

> The PEP proposes exposing the following integer-related functions [1]_ in a new

Reviewer: No, "moving" is a correct word.  We move functions, period.

Author: We're not proposing moving the functions though.  We're proposing leaving them in `math` and _additionally exposing_ them in `imath`.  "Moving" would mean removing from `math` and adding to `imath`.

Are you talking about the implementations?

Backwards Compatibility
=======================

As aliases in :external+py3.14:mod:`math` will be kept indefinitely, there are
no anticipated code breaks.  (The use of these aliases is discouraged.)

Rationale
=========

Why not simply update the documentation to broaden math's scope
---------------------------------------------------------------

Broadening ``math``'s scope would solve the documentation issue, but it would
neglect the motivation of providing a shared context.  As mentioned, shared
context reduces verbosity in the documentation and conceptual load.  It also
aids discoverability through grouping related functions and IDEs suggesting

> helpful completions.

Reviewer: For me it looks like you repeat stuff from the "Motivation" section.  I hope, that with a small improvement of the former we can just remove this section.

Author: Oh, I'm happy to remove the Rationale.  I thought you wanted it.  Should I remove it now?

Module name
-----------

There is already an ``imath`` project on PyPI, but only with two releases

> (0.0.1 & 0.0.2), most recently 4 years ago in 2021.

Reviewer: ```suggestion
There is already an ``imath`` project on PyPI, but only with two releases,
with most recent one 4 years ago.  It's repository not accessible anymore.
```

Author: Done.

Polling shows ``intmath`` as another popular name.  The argument made was that
the normal mathematical spelling of the imaginary unit is i, which makes imath
ambiguous.  It also has no conflict with any PyPI module.  On the other hand,
perhaps ``intmath`` may be confused with interval math or numerical
integration.

Other proposed names include ``ntheory`` (like SymPy's submodule),
``integermath`` and ``imaths``.

Module scope and possible library members
=========================================

> The following is a list of functions that could one day find a home in the
> ``imath`` library.  New fucntions are not part of the proposal.  They
> illustrate the proposed scope of ``imath`` and its potential utility.
> Functions proposed in the initial discussion thread include:
> 
> * Divisibility & Modular Arithmetic (see also [5]_)
> 
>   * ``gcd(a, b)``: Already in ``math``.
>   * ``lcm(a, b)``: Already in ``math``.
>   * ``powmod(a, b, m)``: Modular exponentiation (already in pow(a, b, m), but
>     renamed for clarity).
>   * ``is_coprime(a, b)``
>   * ``modinv(a, m)``: Modular inverse (or perhaps better yet, ``gcdext``
>     below).
>   * ``gcdext()`` --- to solve linear Diophantine equation in two variables (the
>     :external+py3.14:class:`int` implementation actually includes extended
>     Euclidean algorithm).
> 
> * Integer Arithmetic
> 
>   * ``ceil_div(a, b)``: Ceiling divide, see [2]_, [3]_.
>   * ``ilog(n, base)``: Integer logarihm; currently
>     :external+py3.14:func:`math.log` has a special handling for integer
>     arguments.  It's unique (wrt other module functions) and not documented so
>     far. See [4]_.
> 
> * Combinatorics and Integer Sequences
> 
>   * ``factorial(n)``: Already in ``math``.
>   * ``comb(n, k)``: Already in ``math``.
>   * ``fibonacci(n)``
> 
> * Integer Roots
> 
>   * ``isqrt(n)``: Already in ``math``.
>   * ``isqrt_rem(n)``: The integer square root and its remainder.
>   * ``sqrtmod(n, modulo)``: Square root within a modular group.
> 
> * Primality & Factorization
> 
>   * ``is_prime(n)``: Optimized for large numbers.
>   * ``prime_factors(n)``: Trial division or Pollard’s rho for small inputs.
>   * ``primes_between(lower, upper)``: Uses the sieve of Atkin for large ranges
>     that start near zero, or an iterative algorithm for small ranges that are
>     far from zero.
>   * ``totient(n)``: Euler’s totient function.
>   * ``next_prime(n)``
>   * ``prev_prime(n)``
> 
> * Encoding & Number Representations
> 
>   * ``gray_encode(n), gray_decode(n)``
>   * ``hamming_weight(n)``
>   * ``hamming_distance(a, b)``
> 
> * Binary Arithmetic
> 
>   * ``is_power_of_two(x)``
>   * ``next_power_of_two(x)``

Reviewer: ```suggestion
To illustrate the proposed scope of ``imath`` and its potential utility
we list here some functions from the initial discussion thread (see also [5]_),
that could one day find a home in the new module:

* ``c_div()`` --- for integer ceiling divide, see [2]_, [3]_.
* ``gcdext()`` --- to solve linear Diophantine equation in two variables (the
  :external+py3.14:class:`int` implementation actually include extended
  Euclidean algorithm)
* ``isqrt_rem()`` --- to return both integer square root and a remainder (if
  integer isn't a perfect square)
* ``ilog()`` --- integer logarithm, currently :external+py3.14:func:`math.log`
  has a special handling for integer arguments.  It's unique (wrt other module
  functions) and not documented so far, see [4]_
* ``fibonacci()``.

New functions *are not* part of the proposal.  But we should note
that, unless we can just provide bindings to some well supported mathematical
library like the GMP --- the module scope should be limited to leave
really advanced mathematics to external packages like the
`gmpy2 <https://pypi.org/project/gmpy2/>`_.  For example, no
primality testing and factorization of integers.
```
I don't think we should be exhaustive here.  Lets just provide few examples.  What's important here, I think, it's to emphasize that we are not open the door for a feature creep in the stdlib and not propose to include really advanced mathematical functions.

BTW recently I did a simple wrapper of the GMP (https://github.com/diofant/python-gmp).  To be able to run mpmath/diofant on top of that, module was extended to include most functions from the above list.

Author: We're not proposing anything.  The point of the list is to:
* illustrate the core topics of things _might_ fit into the module,
* to support those core topics with potential examples, and
* to negate any arguments that the module is too niche.

That's why I think we should prefer to show more rather than less.  If we just give a handful of functions without organization we lose the core topics.  If we give topics without example functions, the topics are less clear.

And by giving a lot of examples, we negate the arguments against this being a niche module.  I think the strongest argument against `imath` will be from people who say: "Why do we need a module for five functions?"

I've reorganized the topics to _match the abstract_, and support the wording there.  And I've altered the wording to make it clear what this section tries to accomplish.   Let me know what you think :smile:   Maybe we could make it even more clear?  Or reiterate after the list why the list is provided?

Reviewer: > illustrate the core topics of things might fit into the module

I think that plain list do same.  Nobody knows why `isqrt()` and `ilog()` live in different sections.

> to support those core topics with potential examples, and

ditto

> to negate any arguments that the module is too niche.

I doubt it's a severe argument.  E.g. the fraction module has one public entity.

> That's why I think we should prefer to show more rather than less.

Unfortunately, your list miss examples of that is *outside* of the module scope.

Author: > I think that plain list do same.

I don't think a lit of 5 functions gives much indication of the potential scope of the library, and they don't highlight topics either.

> Nobody knows why isqrt() and ilog() live in different sections.

Yes, that makes sense.  I'm perfectly happy recategorizing things if you want.

> I doubt it's a severe argument. E.g. the fraction module has one public entity.

Well, we don't know what will happen.  We want to be defensive?  You may not get a second shot at making all the arguments you want.

> Unfortunately, your list miss examples of that is outside of the module scope.

Are you saying that you want a lit of things that are out of scope?

Happy to just trim this down if you want.  Why don't you decide which lines you want to remove.   I do think we should keep a set of headings though to illustrate scope?

Reviewer: > I don't think a lit of 5 functions gives much indication of the potential scope of the library, and they don't highlight topics either.

We can invent a lot of functions, that map Z->Z.  It's not a problem.

But I think that this section should list few functions, that are suitable for the Python stdlib.  I see following criteria:
1. it's something too common in applications
2. can be efficiently implemented without sophisticated algorithms
3. have not too much redundancy (e.g. is_coprime could be implemented via gcdext, do we need both methods?)

> Are you saying that you want a lit of things that are out of scope?

Sure.  The stdlib can have only tiny subset of number-theoretic functions.

Author: > But I think that this section should list few functions, that are suitable for the Python stdlib.

Okay, we'll do it your way.  I wasn't trying to decide what might be suitable for the stdlib today.  That's a future decision for someone else.  The only point of the list was to establish the scope of what would belong in imath (so having at least a few in each category), and to establish its utility (so the more functions the better).

But no worries, let's do it how you want it.

> 1. it's something too common in applications
> 2. can be efficiently implemented without sophisticated algorithms
> 3. have not too much redundancy (e.g. is_coprime could be implemented via gcdext, do we need both methods?)

Of course this makes sense, but I wasn't thinking about any of these things.  I just wanted to show scope and utility.  But I see your point that it may scare people if the proposal seems larger than it is, so I'm perfectly happy to narrow this.

But, I'm not sure why you're against something like `is_prime`.  That comes up all the time for me when I'm solving advanced puzzles.  And it's easily implemented.

It feels like you think that we're proposing adding these functions and trying to decide what should be added?  I think we should just leave that other people and just propose things that people **_might want to propose_**.

> Sure. The stdlib can have only tiny subset of number-theoretic functions.

We want combinatorics too, right?

Author: Per your suggestions, I threw out integer representations throughout the proposal.

Rejected ideas
==============

> There was a brief discussion about exposing ``math.isqrt`` as ``imath.sqrt``.
> However, ``isqrt`` is ultimately a different function: it is the floor of the
> square root.  It would be confusing to give it the same name (under a different
> module).

Reviewer: ```suggestion
There was a brief discussion about exposing :external+py3.14:func:`math.isqrt`
as ``imath.sqrt``.  However, ``isqrt()`` is ultimately a different function:
it is the floor of the square root.  It would be confusing to give it the same name
(under a different module) as :external+py3.14:func:`math.sqrt` or
:external+py3.14:func:`cmath.sqrt`.
```

Author: Done.

> Arguments against

Reviewer: I think this should be re-worked.

Perhaps, we could add the Rationale section to discuss why proposed module is better than alternatives like more verbose docs, etc.

Author: The arguments against section is just a transcription of what people said about why they don't like this proposal.  The idea is to allow the SC to consider these arguments and to provide us a place to rebut them.

I agree with you that we should rebut the arguments, but I think we should do it right here rather than in another section.  Our rationale is already given in the motivation section where we are as focused as possible on our perspective. 

What do you think?

Reviewer: I don't think we should just literally translate arguments against.

> Our rationale is already given in the motivation section

Per PEP 12:
* Motivation - Clearly explain why the existing language specification is inadequate to address the problem that the PEP solves.
* Rationale - why particular design decisions were made

So, I think that this section is relevant to refute arguments like "why not just fix docs?"  You can take look on other PEPs to ensure.

Author: > So, I think that this section is relevant to refute arguments like "why not just fix docs?"

Yes, you're right that makes sense.  Let's add a _rationale_ section and address this there.

>  I don't think we should just literally translate arguments against.

I agree.  I just didn't want to neglect mentioning counterarguments that were made in the thread.  I always find it a bit dishonest when people do that.  So I wanted to make sure that we at least mentioned everything.

=================

Discoverability
---------------

Beginners looking for a math function may instinctively look for it in the math
module.  Rather than creating a new module, change the documentation for the
math module to read not mention the C standard or floating-point numbers.

Countering this, complex math is still math, but we have a separate module for
that.  Just as with complex numbers, integers should have their own module too.

Limited benefit
---------------

Whereas the ``cmath`` module is justified by its prevention of a lot of slow

> type checks, the ``imath`` module has limited benefit in this regard.  The

Reviewer: I doubt it's the *only* reason why the cmath module exists.

Author: Yes, I'm just transcribing someone's argument!

Other proposed names include ``ntheory`` (like SymPy's submodule),
``integermath`` and ``imaths``.

Module scope and possible library members

> =========================================

Reviewer: I think this should be part of the "Open Issues" section.

Author: I think it's not an open issue because we don't want the SC deciding which library members will be added.  This is just a way of illustrating scope.  We're not looking to start a debate about the choices in this section.

Open issues are for things that the SC needs to decide:  Like the name of the library.

This PEP proposes a new module with only the six existing functions listed
above.  However, in order to support the module's possible future utility, and
to illustrate its scope, we provide an exploratory list of functions that could

> one day find a home in the ``imath`` library:

Reviewer: Sorry, but you again miss my point.

The purpose of this section is to show what can be included eventually to the module and what can't, not classification of the `factorial` as part of the Combinatorics or whatever else.

Author: > classification of the factorial as part of the Combinatorics or whatever else.

But I'm not suggesting the point is classification?  Let's go point-by-point:

Do you agree that this section should:

* show the module's possible future utility (that there are a lot of functions that could find a home here)?
* illustrate the module's scope (what kinds of things would best be in `imath` and not, say, `imath`?

Also, you want his section to show "what can be included eventually".  What do you mean by this exactly?  Is it one of the two things above, or something else?

      but renamed for clarity).
    * ``is_coprime(a, b)``
    * ``modinv(a, m)``: Modular inverse.
    * ``gcdext(a, b)``: Solve linear Diophantine equations in two variables (the
      :external+py3.14:class:`int` implementation actually includes the extended
      Euclidean algorithm).

  * Integer Arithmetic and Sequences

    * ``isqrt(n)``: Part of proposal.
    * ``c_div(a, b)``: Ceiling divide, see [2]_, [3]_.
    * ``isqrt_rem(n)``: The integer square root and its remainder.
    * ``sqrtmod(n, modulo)``: Square root within a modular group.
    * ``ilog(n, base)``: Integer logarihm; currently
      :external+py3.14:func:`math.log` has a special handling for integer
      arguments.  It's unique (wrt other module functions) and undocumented.
      See [4]_.
    * ``fibonacci(n)``

  * Primality & Factorization

    * ``is_prime(n)``
    * ``is_power_of_two(x)``
    * ``prime_factors(n)``: Trial division or Pollard’s rho for small inputs.
    * ``totient(n)``: Euler’s totient function.

* Combinatorics

  * ``factorial(n)``: Part of proposal.
  * ``comb(n, k)``: Part of proposal.
  * ``perm(n, k)``: Part of proposal.

> Rejected ideas
> ==============
> 
> There was a brief discussion about exposing :external+py3.14:func:`math.isqrt`
> as ``imath.sqrt`` in the same way that :external+py3.14:func:`cmath.sqrt` is the
> complex version of :external+py3.14:func:`math.sqrt`. However, ``isqrt`` is
> ultimately a different function: it is the floor of the square root.  It would
> be confusing to give it the same name (under a different module).
> 
> 

Reviewer: ```suggestion
```
There is a copy of that, below.

Author: Okay, I'll remove.

Arguments against
=================

Discoverability
---------------

Beginners looking for a math function may instinctively look for it in the math
module.  Rather than creating a new module, change the documentation for the
math module to read not mention the C standard or floating-point numbers.

Countering this, complex math is still math, but we have a separate module for
that.  Just as with complex numbers, integers should have their own module too.

Limited benefit

> ---------------

Reviewer: I don't see counter arguments here.

Author: Should we add one?  I thought that our counterargument is that it is useful enough for all the reasons we've already given.  Or should we remove this section?  Happy with whatever.

Other proposed names include ``ntheory`` (like SymPy's submodule),
``integermath`` and ``imaths``.

> =======

Reviewer: ```suggestion

```

.. [3] https://gmpy2.readthedocs.io/en/stable/mpz.html#gmpy2.c_div
.. [4] https://github.com/python/cpython/issues/120950
.. [5] https://github.com/python/cpython/issues/81313

> .. [6] :pep:`387`

Reviewer: ```suggestion
```
I think it's redundant.  PEP aren't docs and we already have refrence to the Glossary.

Author: Okay, I'll remove the link!
